utility/sourcemap_accessor.go

```diff
@@ -0,0 +1,192 @@
+	ParseError  Enum = "ParseError"
+)
+
+type SmapError struct {
+	Msg  string
+	Kind Enum
+}
+
+func (e SmapError) Error() string {
+	return fmt.Sprintf("%s", e.Msg)
+}
+
+type SmapID struct {
+	ServiceName    string
+	ServiceVersion string
+	Path           string
+}
+
+type SourcemapAccessor struct {
+	esClients []elasticsearch.Client
+	smapCache *lru.Cache
+}
+
+func NewSourcemapAccessor(esClients []elasticsearch.Client, cacheSize int) (*SourcemapAccessor, error) {
+	if esClients == nil || len(esClients) == 0 {
+		err := SmapError{Msg: "Sourcemap ESClient cannot be initialized.", Kind: InitError}
+		logp.Err(err.Error())
+		return nil, err
+	}
+
+	smapCache, err := lru.New(cacheSize)
+	if err != nil {
+		err := SmapError{Msg: fmt.Sprintf("Sourcemap cache cannot be initialized %s", err.Error()), Kind: InitError}
+		logp.Err(err.Error())
+		return nil, err
+	}
+	return &SourcemapAccessor{esClients: esClients, smapCache: smapCache}, nil
+}
+
+func (s *SourcemapAccessor) Fetch(smapId SmapID) (*sourcemap.Consumer, error) {
+	smapConsumer, err := s.fetchFromCache(smapId)
+	if err != nil {
+		logp.Err(fmt.Sprintf("Sourcemap fetching from Cache Error %s", err.Error()))
+	}
+	if smapConsumer == nil {
+		smapConsumer, err = s.fetchFromES(smapId)
+		if err != nil {
+			logp.Err(fmt.Sprintf("Sourcemap fetching from ES Error %s", err.Error()))
+			return nil, err
+		}
+		s.AddToCache(smapId, smapConsumer)
+	}
+	return smapConsumer, err
+}
+
+func (s *SourcemapAccessor) AddToCache(smapId SmapID, smap *sourcemap.Consumer) {
+	s.smapCache.Add(smapId.key(), smap)
+}
+
+func (s *SourcemapAccessor) RemoveFromCache(smapId SmapID) {
+	s.smapCache.Remove(smapId.key())
+}
+
+func (s *SourcemapAccessor) fetchFromCache(smapId SmapID) (*sourcemap.Consumer, error) {
+	if s.smapCache == nil {
+		msg := "SourceMap Cache requested but not initialized"
+		return nil, SmapError{Msg: msg, Kind: InitError}
+	}
+	if cached, found := (*s.smapCache).Get(smapId.key()); found {
+		return cached.(*sourcemap.Consumer), nil
+	}
+	return nil, nil
+}
+
+func (s *SourcemapAccessor) fetchFromES(smapId SmapID) (*sourcemap.Consumer, error) {
+	if s.esClients == nil || len(s.esClients) == 0 {
+		msg := "Sourcemap ESClient requested but not initialized."
+		return nil, SmapError{Msg: msg, Kind: InitError}
+	}
+
+	body := map[string]interface{}{
+		"query": map[string]interface{}{
+			"bool": map[string]interface{}{
+				"must": []map[string]interface{}{
+					{"term": map[string]interface{}{"processor.name": "sourcemap"}},
+					{"term": map[string]interface{}{"sourcemap.bundle_filepath": smapId.Path}},
+					{"term": map[string]interface{}{"sourcemap.service.name": smapId.ServiceName}},
+					{"term": map[string]interface{}{"sourcemap.service.version": smapId.ServiceVersion}},
+				},
+			},
+		},
+		"size": 1,
+		"sort": []map[string]interface{}{
+			{
+				"@timestamp": map[string]interface{}{
+					"order": "desc",
+				},
+			},
+		},
+	}
+
+	index := "apm-*"
```

Comment:
needs to be configurable?

Reply:
true, I will change to pass in the index with the config.